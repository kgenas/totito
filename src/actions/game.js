import { types } from "../types/types";

export const startGame = () =>({
    type: types.start
})

export const Clicked = (cellValues,numberOfTurn,isNext,isGameOver,winner,winningCombination) =>({
    type: types.cellValues,
    payload:{
        cellValues,
        numberOfTurn,
        isNext,
        isGameOver,
        winner,
        winningCombination
    }  
})

export const NewGame = () =>({
    type: types.newGame
})