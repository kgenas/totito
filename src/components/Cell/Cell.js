import './Cell.css';
import classNames from 'classnames';

export const Cell = ( props ) => {
  
    const cellCss = classNames({
        cell: true,
        winner: props.switchOnOff
    })

    const cellContentCss = classNames({
        'cell--contet': true,
        populated: props.value 
    })

  return (
    <button
        className={ cellCss }
        onClick = { props.onClick }
    >
        <span
            className={ cellContentCss }
        >
            { props.value }
        </span>
    </button>
  )
}
