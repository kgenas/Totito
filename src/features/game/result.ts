import { createSlice, PayloadAction } from '@reduxjs/toolkit';
import { AppThunk } from '../../app/store';

export interface InitialState {
    isGameOver: boolean;
    winner: ( string | undefined );
    winningCombination: number[];
}
const dataArray: InitialState = {
    isGameOver: false,
    winner: undefined,
    winningCombination:[],
};

export const result = createSlice({
  name: 'result',
  initialState:dataArray,  
  reducers: {
    
    winnerCalculator: (state, action: PayloadAction<InitialState>) => { 
      const { isGameOver , winner , winningCombination } = action.payload;
      return {
        ...state,
        isGameOver,
        winner,
        winningCombination
      };
    },  
    winnerCalculatorStart: (state) => {
        state.isGameOver = dataArray.isGameOver;
        state.winner = dataArray.winner;
        state.winningCombination = dataArray.winningCombination;
      },
  },
});

export const { winnerCalculator, winnerCalculatorStart } = result.actions;

  const winnningMatrix:any = {
      0:[[1,2],[3,6],[4,8]], //? example 0,1,2 o 0,3,6 o 0,4,8
      1:[[0,2],[4,7]],
      2:[[0,1],[5,8],[4,6]],
      3:[[0,6],[4,5]],
      4:[[2,6],[3,5],[1,7],[0,8]],
      5:[[3,4],[2,8]],
      6:[[7,8],[0,3],[2,4]],
      7:[[6,8],[1,4]],
      8:[[6,7],[2,5],[0,4]]
  }

  export const winnerCalculatorValue = 
  (cellValues:string[], numberOfTurn:number, cellIndex:number):AppThunk =>
  (dispatch) =>
  {

      if( numberOfTurn === 0 ){
          dispatch(winnerCalculator({
              isGameOver:true,
              winner:undefined,
              winningCombination:[]
          }));        
      };   
      const WinningOptions = winnningMatrix[cellIndex];
      for(let WinningOption of WinningOptions){
          
          const currentValue = cellValues[cellIndex];
          const firstOption =  cellValues[WinningOption[0]];
          const secondOption = cellValues[WinningOption[1]];

          if(currentValue === firstOption && firstOption === secondOption ){             
              dispatch(winnerCalculator({
                  isGameOver:true,
                  winner:currentValue,
                  winningCombination:[cellIndex,WinningOption[0],WinningOption[1]]
              }));            
          };
      };         
  }


export default result.reducer;
