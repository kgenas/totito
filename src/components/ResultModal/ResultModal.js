import classNames from 'classnames';
import { useDispatch, useSelector } from 'react-redux';
import { NewGame } from '../../actions/game';
import './ResultModal.css';

export const ResultModal = () =>{

    const dispatch = useDispatch();
    const { isGameOver, winner } = useSelector( state => state.GameTotito) ;
    const resultMondalClasses = classNames({
        'modal-open': isGameOver
    })
    const message = winner ? `Winner is ${ winner }.` :'It is a tie';

    const newGame = () =>{
        dispatch(NewGame());
    }
    return(
        <div id='modal-overlay' className={ resultMondalClasses }>
            <div id='game-result-modal'>
                <div id='result-container'>
                    <div id='winner-container'>
                        <span>{ message }</span>
                    </div>
                </div>
                <div id='new-game-container'>
                    <button
                        id='new-game-container'
                        onClick={ newGame }
                    >
                        Start New Game
                    </button>
                </div>
            </div>            
        </div>
    )
}