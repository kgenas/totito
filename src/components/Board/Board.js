import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell/Cell';

export const Board = (props) => {
  
  const { cellValues, winningCombination } = useSelector( state => state.Game)
  const cell = cellValues.map((value, index)=>{
        
    const switchOnOff = winningCombination && winningCombination.indexOf(index) >= 0;
    return(
        <Cell
            key= { index }
            value = { value }
            switchOnOff = { switchOnOff }
            onClick = { () => props.cellClicked(index) }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
