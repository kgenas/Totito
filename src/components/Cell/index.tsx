import './Cell.css';
import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorValue } from '../../features/game/result';
import { click } from '../../features/game/totito';

interface valueProps {    
    key: number ;
    value: string;
    valueKey: number;
    switchOnOff: boolean;
}
export const Cell = ( props: valueProps ) => {
          
    const { cellValues, numberOfTurn, xIsNext } = useSelector((state: RootState) => state.game );    
    const dispatch:AppDispatch = useDispatch();

    const isCellEmpty = ( cellIndex:number ) => cellValues[cellIndex] === '';
    const onCellClicked = () =>{ 
      
      if (isCellEmpty( props.valueKey )){
        const newCellValues = [...cellValues];
        newCellValues[props.valueKey] = xIsNext ? 'X':'O';
        dispatch(winnerCalculatorValue( newCellValues,(numberOfTurn-1),props.valueKey ));        
        dispatch(click(
            {
                cellValues: newCellValues,
                numberOfTurn: (numberOfTurn-1),
                xIsNext: !xIsNext,
            }
        ));
    }
    }
    const cellCss = classNames({
        cell: true,
        winner: props.switchOnOff
    })

    const cellContentCss = classNames({
        'cell--contet': true,
        populated: props.value 
    })

  return (
    <button
        className={ cellCss }
        onClick = { onCellClicked }
    >
        <span
            className={ cellContentCss }
        >
            { props.value }
        </span>
    </button>
  )
}
