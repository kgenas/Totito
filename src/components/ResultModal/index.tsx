import classNames from 'classnames';
import { useSelector, useDispatch } from 'react-redux';
import { newGame } from '../../features/game/totito';
import './ResultModal.css';
import { AppDispatch, RootState } from '../../app/store';
import { winnerCalculatorStart } from '../../features/game/result';

export const ResultModal = () =>{

    const dispatch:AppDispatch = useDispatch();
    const { winner , isGameOver } = useSelector((state: RootState) => state.result );    

    const resultMondalClasses = classNames({
        'modal-overlay': true,
        'modal-overlay--modal-open': isGameOver

    })
    const message = winner ? `Winner is ${ winner }.` :'It is a tie';

    const newGames = () =>{
        dispatch(newGame());
        dispatch(winnerCalculatorStart());
    }
    return(
        <div className= { resultMondalClasses }>
            <div className='game-result-modal'>
                <div className='result-container'>
                    <div className='winner-container'>
                        <span>{ message }</span>
                    </div>
                </div>
                <div className='new-game-container'>
                    <button
                        className='new-game-container__button'
                        onClick={ newGames }
                    >
                        Start New Game
                    </button>
                </div>
            </div>            
        </div>
    )
}