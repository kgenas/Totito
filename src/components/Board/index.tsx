import './Board.css';
import { useSelector } from 'react-redux';
import { Cell } from '../Cell';
import { RootState } from '../../app/store';
export const Board = () => {
    
  const { cellValues }  = useSelector((state: RootState) => state.game );
  const { winningCombination } = useSelector((state: RootState) => state.result ); 

  const cell = cellValues.map((value:string, index:number)=>{
       
  const switchOnOff = winningCombination && winningCombination.indexOf( index ) >= 0;
    return(
        <Cell
            key = { index }
            value = { value }
            valueKey = { index }
            switchOnOff = { switchOnOff }
        />
    )
  })

  return (
    <div className='board'>
        { cell }
    </div>
  )
}
