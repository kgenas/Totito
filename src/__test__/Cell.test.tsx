/* eslint-disable testing-library/no-debugging-utils */
import React from 'react';
import { screen } from '@testing-library/react';
import { setupStore } from '../app/store';
import { Cell } from '../components/Cell/index';
import { renderWithProviders } from '../utils/test-utils';
import { click } from '../features/game/totito';
import { Board } from '../components/Board';
import { winnerCalculatorValue } from '../features/game/result';
import { ResultModal } from '../components/ResultModal';

describe('Cell',() =>{

    // Note: running cleanup afterEach is done automatically for you in @testing-library/react@9.0.0 or higher
    // unmount and cleanup DOM after the test is finished.
    //? afterEach(cleanup);
    test('Renders Cell component',() => {
        renderWithProviders(<Cell key={1} value={''} valueKey={0} switchOnOff={false} />);        
    });
    test('Matches snapshot',() => {
        
        const { baseElement } = renderWithProviders(<Cell key={2} value={''} valueKey={2} switchOnOff={false} />);
        expect( baseElement ).toMatchSnapshot();
    });

    test('Click Cell', () => {        
        const store = setupStore();
        const newCellValues:string[] = [];
        newCellValues[0] = 'X';        
        store.dispatch(click({
                    cellValues:newCellValues,
                    numberOfTurn: (9-1),
                    xIsNext: false,
        }))
        renderWithProviders(<Board/>, { store });
        const value = screen.getAllByRole('button');
        expect(value[0]).toHaveTextContent('X');        
        screen.debug();
     });
     
    test('Winner X Calculator', () => { 
        const store = setupStore();
        const newCellValues:string[] = ['X','X','X']; //* 0,1,2

        store.dispatch(winnerCalculatorValue(newCellValues,(5-1),0,));
        renderWithProviders(<ResultModal/>, { store })        
        expect(screen.getByText(/Winner is X./i)).toBeInTheDocument();
        screen.debug();
    });
});

