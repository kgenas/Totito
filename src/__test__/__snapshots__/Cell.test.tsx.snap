// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cell Matches snapshot 1`] = `
<body>
  <div>
    <button
      class="cell"
    >
      <span
        class="cell--contet"
      />
    </button>
  </div>
</body>
`;
